src/awsme/buffered_recorder.py

```diff
@@ -20,6 +21,7 @@ class BufferedRecorder:
     def __init__(self, *args, recorder=None, **kwargs) -> None:
         self._recorder = recorder or ImmediateRecorder(*args, **kwargs)
         self._buffer = []  # type: List[Dict[str, Any]]
+        atexit.register(self.flush) # ensure that we flush before we depart this earth
```

Comment:
I wonder if this can be unit tested. Could you please take look at it?

Reply:
Yes, I'll look at it now

Reply:
OK - I think that it can be tested, with some effort. Basically, we have to:

- Create a new recorder that outputs something to STDOUT, or update the DummyRecorder to do the same
- Create another test .py that creates a buffered recorder, using the new recorder above on input, and puts one metric
- Start another interpreter with subprocess.run(), running the new .py test file created above
- Assert that the CompletedProcess returned from run() has a stdout result that matches our expectations

I don't have the time to write that ATM, but if you create an issue I'm happy to pick it up in the near future and add the test...

-Joe

Reply:
Plan sounds good. But I think it makes sense to put the while setup into a separate file and execute it using exec, instead of subprocess. No need to modify Dummy recorder or anything else in the library itself.
No rush, the test should go into the same PR. Then I’ll be confident in merging.

Reply:
I'm a little confused. Without catching verifiable output on STDOUT, how are we to ensure that the atexit is actually doing anything? And my reading of the current recorders doesn't put anything to STDOUT - or did I miss something?

I don't think that exec will work, unless once again I'm missing something. My understanding is that for functions registered by atexit to run, the interpreter must be terminating. I don't think that exec does that and will thus not trigger atexit.

I'll figure it out...

Reply:
Here are atexit tests in CPython:
https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b
/Lib/test/test_atexit.py#L213

I mean, put the new STDOUT recorder right inside the file, that will be `exec`ed.

Reply:
My bad, you're right, they don't use `exec`.

Reply:
I'm working on it now. Should have something to you in a couple of hours.

Reply:
Test added, test passed

Reply:
Awesome!